packages/block-editor/src/components/link-control/index.js

```diff
@@ -220,21 +299,60 @@ function LinkControl( {
 		setIsEditingLink( false );
 	}
 
+	/**
+	 * Determines whether a given value could be a URL. Note this does not
+	 * guarantee the value is a URL only that it looks like it might be one
+	 * (thus the naming of the method).
+	 *
+	 * @param {string} val the candidate for being a valid URL (or not).
+	 */
+	function maybeURL( val ) {
+		const isInternal = startsWith( val, '#' );
+
+		return isInternal || isURL( val ) || ( val && val.includes( 'www.' ) );
+	}
+
 	// Effects
 	const getSearchHandler = useCallback(
 		( val, args ) => {
-			const isInternal = startsWith( val, '#' );
-
-			const handleManualEntry =
-				isInternal || isURL( val ) || ( val && val.includes( 'www.' ) );
-
-			return handleManualEntry
+			return maybeURL( val )
 				? handleDirectEntry( val, args )
 				: handleEntitySearch( val, args );
 		},
 		[ handleDirectEntry, fetchSearchSuggestions ]
 	);
 
+	const handleOnCreate = async ( suggestionTitle ) => {
+		setIsResolvingLink( true );
+		setErrorMessage( null );
+
+		// Make cancellable in order that we can avoid setting State
+		// if the component unmounts during the call to `createSuggestion`
+		cancelableCreateSuggestion.current = makeCancelable(
+			// Using Promise.resolve to allow createSuggestion to return a
+			// non-Promise based value.
+			Promise.resolve( createSuggestion( suggestionTitle ) )
+		);
+		const newSuggestion = await cancelableCreateSuggestion.current.promise;
+
+		// ********
+		// NOTE: if the above Promise rejects then code below here will never run
+		// ********
+		setIsResolvingLink( false );
+
+		// Only set link if request is resolved, otherwise enable edit mode.
+		if ( newSuggestion ) {
+			onChange( newSuggestion );
+		} else {
+			setIsEditingLink( true );
+		}
+	};
+
+	const handleSelectSuggestion = ( suggestion, _value = {} ) => {
+		setIsEditingLink( false );
+		onChange( { ..._value, ...suggestion } );
```

Comment:
We're perpetuating this problem and making it harder to fix: #20020

Can you review my pull request?

Reply:
I can indeed. Let me look.

Reply:
> We're perpetuating this problem

You're referring to the fact we're duplicating the code merging `value` and `suggestion`?

I [tried reviewing but hit roadblocks](https://github.com/WordPress/gutenberg/pull/20020#issuecomment-586968911) and I don't fully follow the issue.

Reply:
Followed up on https://github.com/WordPress/gutenberg/pull/20020. We'll take it from there.